internal/tcprunner/tcprunner.go

```diff
@@ -0,0 +1,192 @@
+package tcprunner
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"net"
+	"strings"
+
+	"github.com/ooni/probe-cli/v3/internal/measurexlite"
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/tracex"
+)
+
+// Model describes a type that does a DNS lookup(s), then attempts several TCP sessions
+type Model interface {
+	// Stores the provided hostname
+	Hostname(string)
+	// Store DNS query result
+	DNSResults([]*model.ArchivalDNSLookupResult)
+	// Indicates one or more steps failed (can be overwritten)
+	Failed(string)
+	// Stores a new individual test key (for a TCP session) and returns a pointer to it
+	NewRun(string, string) TCPSessionModel
+}
+
+// TCPSessionModel describes a type that does a single TCP connection and TLS handshake with a given IP/Port combo
+type TCPSessionModel interface {
+	// Store IP/port address used for this session
+	IPPort(string, string)
+	// Store TCP connect result
+	ConnectResults([]*model.ArchivalTCPConnectResult)
+	// Store TLS handshake result
+	HandshakeResult(*model.ArchivalTLSOrQUICHandshakeResult)
+	// Indicates a failure string, as well as an identifier for the failed step
+	FailedStep(string, string)
+}
+
+// TCPRunner manages sequential TCP sessions to the same hostname (over different IPs)
+type TCPRunner struct {
+	Tk        Model
+	Trace     *measurexlite.Trace
+	Logger    model.Logger
+	Ctx       context.Context
+	Tlsconfig *tls.Config
+}
+
+// TCPSession Manages a single TCP session and TLS handshake to a given ip:port
+type TCPSession struct {
+	Itk     TCPSessionModel
+	Runner  *TCPRunner
+	Addr    string
+	Port    string
+	TLS     bool
+	RawConn *net.Conn
+	TLSConn *net.Conn
```

Comment:
We discussed out of band the issue about the deadlock in the startTLS method. While it's still unclear to me exactly what the root cause is, one problem in the current code is that these `RawConn` and `TLSConn` should not be pointers (`net.Conn` is an interface), but an actual copy to the relevant connection.

`net.Conn` makes some guarantees about it being possible to call methods on them also concurrently, which means it has in it some locking to deal with concurrency and is probably what is leading to the deadlock.

I haven't fully understood where exactly this is happening, but I suspect it has something to do with the fact that there is this proxy object managing the two connections.

My suggestion would be to refactor the code to get rid of this proxy object and use the `net.Conn` objects directly. You don't even need to handle the special case for startTLS, as you can conditionally perform startTLS on the plaintext connection and swap it out for the original `net.Conn` once you have an encrypted tunnel.

That should simplify the code and hopefully make it more clear where the deadlock is happening.

Reply:
So i just pushed [a new commit](https://github.com/ooni/probe-cli/pull/989/commits/cd68f3958219ba5f815832b1e159dd27b3bfa7b3) with this suggestion. Unfortunately the code is still deadlocking. strace log:
<pre><code>
[      5.571522] <info> Initializing SMTP client
epoll_pwait(3, [], 128, 0, NULL, 0)     = 0
write(5, "\0", 1)                       = 1
futex(0x2482168, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x2482168, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
epoll_pwait(3, [], 128, 0, NULL, 0)     = 0
futex(0x2482168, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
epoll_pwait(3, [], 128, 0, NULL, 0)     = 0
nanosleep({tv_sec=0, tv_nsec=3000}, NULL) = 0
write(5, "\0", 1)                       = 1
futex(0x2482168, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
epoll_pwait(3, [], 128, 0, NULL, 0)     = 0
futex(0xc000088148, FUTEX_WAKE_PRIVATE, 1) = 1
futex(0x2482168, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x2482168, FUTEX_WAIT_PRIVATE, 0, NULL) = 0
futex(0x2482168, FUTEX_WAIT_PRIVATE, 0, NULL
</code></pre>